mr/src/main/java/org/apache/iceberg/mr/mapred/TableResolver.java

```diff
@@ -0,0 +1,121 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.mr.mapred;
+
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.Properties;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.mapred.JobConf;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.hadoop.HadoopCatalog;
+import org.apache.iceberg.hadoop.HadoopTables;
+import org.apache.iceberg.mr.InputFormatConfig;
+
+final class TableResolver {
+
+  private TableResolver() {
+  }
+
+  static Table resolveTableFromJob(JobConf conf) throws IOException {
+    Properties properties = new Properties();
+    properties.setProperty(InputFormatConfig.CATALOG_NAME, extractProperty(conf, InputFormatConfig.CATALOG_NAME));
+    if (conf.get(InputFormatConfig.CATALOG_NAME).equals(InputFormatConfig.HADOOP_CATALOG)) {
+      properties.setProperty(InputFormatConfig.SNAPSHOT_TABLE, conf.get(InputFormatConfig.SNAPSHOT_TABLE, "true"));
+    }
+    properties.setProperty(InputFormatConfig.TABLE_LOCATION, extractProperty(conf, InputFormatConfig.TABLE_LOCATION));
+    properties.setProperty(InputFormatConfig.TABLE_NAME, extractProperty(conf, InputFormatConfig.TABLE_NAME));
+    return resolveTableFromConfiguration(conf, properties);
+  }
+
+  static Table resolveTableFromConfiguration(Configuration conf, Properties properties) throws IOException {
+    String catalogName = properties.getProperty(InputFormatConfig.CATALOG_NAME);
+    URI tableLocation = pathAsURI(properties.getProperty(InputFormatConfig.TABLE_LOCATION));
+    if (catalogName == null) {
+      throw new IllegalArgumentException("Catalog property: 'iceberg.catalog' not set in JobConf");
+    }
+    switch (catalogName) {
+      case InputFormatConfig.HADOOP_TABLES:
+        HadoopTables tables = new HadoopTables(conf);
+        return tables.load(tableLocation.getPath());
+      case InputFormatConfig.HADOOP_CATALOG:
+        String tableName = properties.getProperty(InputFormatConfig.TABLE_NAME);
+        TableIdentifier id = TableIdentifier.parse(tableName);
+        if (tableName.endsWith(InputFormatConfig.SNAPSHOT_TABLE_SUFFIX)) {
+          if (!Boolean.parseBoolean(properties.getProperty(InputFormatConfig.SNAPSHOT_TABLE,
+                  Boolean.TRUE.toString()))) {
+            String tablePath = id.toString().replaceAll("\\.", "/");
+            URI warehouseLocation = pathAsURI(tableLocation.getPath().replaceAll(tablePath, ""));
+            HadoopCatalog catalog = new HadoopCatalog(conf, warehouseLocation.getPath());
+            return catalog.loadTable(id);
+          } else {
+            return resolveMetadataTable(conf, tableLocation.getPath(), tableName);
+          }
+        } else {
+          URI warehouseLocation = pathAsURI(extractWarehousePath(tableLocation.getPath(), tableName));
```

Comment:
We had initially implemented the IF using a method similar to what you're describing: we required a user to add an extra table property called `iceberg.warehouse.location` I think and used that to resolve the table with HadoopCatalog. The `LOCATION` property of your Hive table has to be unique for each table (which is why we always have a property for a full table path), so we couldn't put a path to a common warehouse there. We initially tried using the same location for multiple tables and relied on the table name/ID to resolve the table correctly in Iceberg but that caused major confusion in Hive when trying to join tables :') 

So originally we used an extra property for the warehouse location, but we realised we could extract this location from the full table path and save the user an extra configuration property. We don't have any strong preference either way, so I'm happy to change this as you've suggested

Although thinking about this raises another point - I've just realised/tested that you could create an Iceberg table using HadoopCatalog but if you had the full path you can use HadoopTables to load the table again instead... And with Hive we can always expect to have the full path to the table because the location needs to be unique, so we could simplify here and only use HadoopTables for this IF? @massdosage

Reply:
IIRC when the InputFormat "runs" there is always a value in the associated HiveConf named "location" (`InputFormatConfig.TABLE_LOCATION` above) which points to the "base" location of the table being queried. So I think we should be able to just use that and require that the location is set correctly when the Hive table is created which is the normal way of doing things for external tables. Then we can just use that here and not worry about additional table properties or the other code paths which hopefully simplifies things.

Reply:
> I've just realised/tested that you could create an Iceberg table using HadoopCatalog but if you had the full path you can use HadoopTables to load the table again instead...

Yes, this is why I said there is no benefit to using HadoopCatalog. If you already have the full path, you can open the table with HadoopTables and not worry about the warehouse path to pass to the catalog.

The only drawback is that this approach doesn't work for Hive tables. You'd need to know whether the table is a Hive table or a Hadoop table because they use different commit mechanisms.

For tables tracked by the iceberg-hive connector, you'd just need to instantiate the HiveCatalog and use the table name to load the table.

Reply:
OK, may I suggest we just focus on supporting HadoopTables using the location for now and then add support for HiveCatalog in a future PR after this is merged? Just to keep the scope of this PR down etc.

Reply:
Yeah, that sounds good.